concurrency/WorldStringDBNameProvider.java

```diff
@@ -0,0 +1,132 @@
+package concurrency;
+
+import amazon.DeviceMaster.service.InvalidArgumentException;
+import amazon.DeviceMaster.service.util.GuiceConfig;
+import amazon.fiona.util.SuperLogger;
+import amazon.platform.config.AppConfig;
+import amazon.platform.profiler.Profiler;
+import amazon.platform.profiler.ProfilerScope;
+import com.amazon.DMS.Localization.LocalizationHelper;
+
+/**
+ * Concrete class for TemplateNameProvider interface. Populates the deviceName and defaultDeviceName
+ * from worldStringDB  
+ * @author sushils
+ *
+ */
+
+public class WorldStringDBNameProvider implements TemplateNameProviderInterface
+{
+    private static final SuperLogger log = SuperLogger.getLogger(WorldStringDBNameProvider.class);
+    private static String DEFAULT_DEVICE_TYPE = "kindle";
+    private String m_deviceName;
+    private String m_defaultDeviceName;
+
+    private String m_deviceType;
+    private String m_language;
+    private String m_defaultDeviceType;
+
+    static LocalizationHelper m_localizationHelper = 
+            GuiceConfig.getInjector().getInstance(LocalizationHelper.class);
+
+    public WorldStringDBNameProvider(final String deviceType, final String language)
+    {
+        if(deviceType != null && deviceType.trim().length() >= 0) {
+            m_deviceType = deviceType;
+        } else {
+            m_deviceType = DEFAULT_DEVICE_TYPE;
+        }
+
+        if(language != null && language.trim().length() >=0 ) {
+            m_language = language;
+        } else {
+            m_language = "en-US";
+        }
+    }
+
+    public void setDefaultDeviceType(final String defaultDeviceType)
+    {
+        if (defaultDeviceType != null && defaultDeviceType.trim().length() >= 0) {
+            m_defaultDeviceType = defaultDeviceType;
+        } else {
+            m_defaultDeviceType = DEFAULT_DEVICE_TYPE;
+        }
+    }
+
+    @Override
+    public String getDeviceName() throws InvalidArgumentException
+    {
+        if(m_deviceName != null) {
+            return m_deviceName;
+        }
+        final String deviceNameTemplate = AppConfig.findString("DeviceNameTemplate");
+        m_deviceName = getLocalizedTemplate(deviceNameTemplate, m_language);
+        return m_deviceName; 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 59:64, 70:75.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.